mc/Mongo-Pharo-Tools.package/Mongo.extension/instance/openWithTimeout..st

```diff
@@ -0,0 +1,10 @@
+*Mongo-Pharo-Tools
+openWithTimeout: aTimeout
+	| addr |
+	"Waits the specified number of seconds to open the connection."
+	self flag: #todo. "TODO(zecke): Make aTimeout a go like Context."
```

Comment:
What's the meaning of the todo note?

Reply:
The first part is that most timeouts in Pharo are inconsistent. Is the type of the argument a Duration, a SmallInteger or a Float? In case of SmallInteger/Float what is the unit (e.g. openWithTimeout: 0.1)?

The second part is an absolute timeout value is not too helpful. Let me give you one example. Let's assume we receive a function call through gRPC. The call has a timeout attached to it (let's say 5 seconds, so provide a result within 5s or the caller considers it failed and will retry/fail anyway).

A first implementation might look like:

```
 objA := self doSomeDatabaseStuff: paramA.
 objB := self doSomeDatabaaseStuff: param B.
 cmd := self buidlCommandWith: objA with: objB.
 ^cmd execute.
```

But that's problematic as this might block the server for more than 5 seconds and resources are used for the canceled operation. It's a classic scenario/behavior in case of overload and amplifies. A second implementation might look like.

```
objA := self doSomeDatabaseStuff: paramA timeout: aTimeout / 3.
objB := self DoSomeDatabaseStuff: paramB timeout: aTimeout / 3.
^ (self buildCommandWith: objA with: objB) executeWithTimeOut: aTimeout /3.
```

Now the static division of timeout isn't nice either. Maybe the first database look-up initialize the db pool and takes longer. Then we fail requests besides having had enough time.

In Go this is solved by passing a "context" to functions. The context is hierarchical and carries/can carry an absolute timeout. So at any time one can query it for how much time is remaining and decide to not execute/return from a call in case of timeout. An example of it is here: https://golang.org/pkg/context/#example_WithDeadline.


I am not convinced that the imperative approach is the nicest for smalltalk but the concept seems desirable to me.

```
TimeoutContext parent: XyzContext current withValue: 5 seconds during: [
   objA := self doSomeDatabaseStuff: paramA.
  objB := self doSomeDatabaaseStuff: param B.
  cmd := self buidlCommandWith: objA with: objB.
  ^cmd execute.
]
``` 

The above (with some form of VM integration for native operations) might be nice.

Reply:
Thanks. 
+1000 to Duration timeouts
So... TimeoutContext sound good.

Reply:
I suggest:
~~~
openWithTimeout: aDuration
	| addr |
	"Waits the specified number of seconds to open the connection."
	self flag: #todo. "TODO(zecke): Make aTimeout a go like Context."
	addr := NetNameResolver addressForName: host timeout: 20.
	stream := SocketStream openConnectionToHost: addr port: port timeout: aDuration asSeconds.
	stream socket setOption: 'TCP_NODELAY' value: 1.
	stream binary.
	authCache := nil
~~~

But the todo may be more useful as an issue ("feature request" or "question" to discuss on it).

The 20 seconds are hardcoded, how should it be better parameterized? an instvar?

Reply:
* Removed the TODO for now
* The 20 seconds are bogus. It's blocking on Unix anyway. We need to get better DNS into the image. ;)

Reply:
I checked if Zinc has some alternative implementation but no: it uses NetNameResolver too.